### PR TITLE
Fix datetime transform

### PR DESCRIPTION
**Pull request by nickpell**
fixes http://manage.dimagi.com/default.asp?182796

@NoahCarnahan @czue

### CODE REVIEW

File: corehq/apps/userreports/expressions/getters.py [hunk 1]
```diff
@@ -75,11 +76,31 @@ def recursive_lookup(dict_object, keys):
...
+            return item
+    return None
 
 
 def transform_datetime(item):
-    return item or None
+    if item:
+        if isinstance(item, basestring):
+            for datetime_format in [
+                '%Y-%m-%dT%H:%M:%SZ',
```
Comment (nickpell) [inline corehq/apps/userreports/expressions/getters.py]: @dannyroberts I followed the example of ```DateTimeProperty``` - do these formats correspond to those in use in HQ?

File: corehq/apps/userreports/expressions/getters.py [hunk 2]
```diff
@@ -75,11 +76,31 @@ def recursive_lookup(dict_object, keys):
 
 def transform_date(item):
     # postgres crashes on empty strings, but is happy to take null dates
-    return item or None
+    if item:
+        if isinstance(item, basestring):
+            try:
+                return datetime.strptime(item, '%Y-%m-%d').date()
```
Comment (nickpell) [inline corehq/apps/userreports/expressions/getters.py]: @dannyroberts same question for dates

File: corehq/apps/userreports/expressions/getters.py (hunk 1 above)
Comment (czue) [inline corehq/apps/userreports/expressions/getters.py]: I think this is the equivalent of `iso_string_to_datetime(item, strict=True)` which might be what you want

File: corehq/apps/userreports/expressions/getters.py (hunk 2 above)
Comment (czue) [inline corehq/apps/userreports/expressions/getters.py]: (and `iso_string_to_date`)

**Comment by nickpell** [comment]
@czue made changes

**Comment by czue** [comment]
@nickpell is this necessary? i think the root cause might have been https://github.com/dimagi/dimagi-utils/pull/321

**Comment by nickpell** [comment]
@czue You can execute the following code on master and on this branch to see the difference:

```
from casexml.apps.case.models import CommCareCase
from corehq.apps.userreports.expressions.factory import ExpressionFactory

expression = ExpressionFactory.from_spec({"datatype": "datetime","type": "property_name","property_name": "opened_on"})
case = CommCareCase.get(case_id) # valid case_id

print expression(case._doc)
print type(expression(case._doc))
```

Even with https://github.com/dimagi/dimagi-utils/pull/321, this change is still necessary to convert ```str``` to ```date```/```datetime```, otherwise additional transforms won't work.

**Comment by czue** [comment]
what is `opened_on` in the database that it's not coming back as a datetime?

File: corehq/apps/userreports/tests/test_app_manager_integration.py [hunk 3]
```diff
@@ -54,12 +54,14 @@ def test_simple_case_management(self, datetime_mock):
         )
         self.assertEqual(expected_columns, set(col_back.id for col_back in data_source.get_columns()))
 
+        modified_on = datetime(2014, 11, 12, 15, 37, 49)
+        opened_on = datetime(2014, 11, 11, 23, 34, 34, 25)
         sample_doc = dict(
             _id='some-doc-id',
             doc_type="CommCareCase",
-            modified_on="2014-11-12T15:37:49",
+            modified_on=modified_on.isoformat() + 'Z',
```
Comment (dannyroberts) [inline corehq/apps/userreports/tests/test_app_manager_integration.py]: I think you want `json_format_datetime` here

**Comment by nickpell** [comment]
```case._doc["opened_on"]``` is a ```str``` (```'2014-12-10T17:26:40.000000Z'```), ```case.opened_on``` is a ```datetime```.  The latter is wrapped as a ```JsonObject``` property, and the former is not.  https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/models.py#L200 takes a ```dict```, not a ```JsonObject```

**Comment by czue** [comment]
thanks - that was the missing piece i was looking for. sorry about all the questions!

**Pull request merged**